### PR TITLE
Add local version extension

### PR DESCRIPTION
**Pull request by drewr**
In cases where we need to publish a new version of our JNA jar apart from upstream, we need a way to identify it from the last.  This implements a scheme similar to the way OS packages do it, where we add an ordinal after the major.minor.patch ("pre-release" in semver).

### CODE REVIEW

**Comment by nik9000** [comment]
I tried to lookup what debian calls these, and, unhelpfully, they call it the ["debian_revision"](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version).

**Comment by rjernst** [comment]
Can we only do this when there is a change actually needed? ie we only add the `-1` suffix when we need to patch a released version.

**Comment by drewr** [comment]
@rjernst Are you proposing just manually patch the code when this needs to be done as opposed to changing the versioning scheme more permanently?

**Comment by rjernst** [comment]
I'm saying the first release for a given "real" jna version would match the jna version. Then, if in this rare scenario we need to modify our jar (eg adding more architectures like in this case), we append "-1", "-2" and so on to the version (that we release). The version in jna would always stay the same, this would just be in our maven artifacts.

**Comment by drewr** [comment]
@rjernst I've added a `prerelease` flag (naming suggestions welcome) that will optionally enable this behavior if needing to publish an alternate to maven central only. The full story is in the [readme](https://github.com/elastic/jna-build/blob/2ea060dd5929f53370d14a428c6f773e5e73299e/README.txt#L4-L49).

**Comment by drewr** [comment]
Actually, [this version of the readme](https://github.com/elastic/jna-build/blob/4ea6d0003211efdbb756d1526aeacc80acfd3587/README.txt#L4-L50) makes more sense.

**Comment by rjernst** [comment]
> I've added a prerelease flag (naming suggestions welcome)

I don't think this has anything to do with prerelease (which to me implies alpha/beta/preview type naming). It's more a "version suffix"?

And thank you for the readme! I think it only works, though, when a suffix is present? I would just have the suffix include the dash.

File: build.gradle [hunk 1]
```diff
@@ -136,6 +144,7 @@ publishing {
         node.appendNode('name', 'Elastic JNA Distribution')
         node.appendNode('description',
           'A build of jna which supports all platforms supported by Elasticsearch')
+        node.appendNode('url', 'https://github.com/elastic/jna-build')
```
Comment (rjernst) [inline build.gradle]: Why is this necessary? I don't think we should use our own repo because we are just repackaging jna. But I was also able to publish this to maven central before without adding it (at least I don't remember modifying the pom manually...).

Comment (drewr) [reply build.gradle]: It kept failing without it when I tried to close the repository.  I can regenerate and take a screenshot.

I thought ours would be more appropriate since we actually packaged it, I'm fine with the upstream. Will change.

**Comment by drewr** [comment]
I'll rework the readme for the more general case and switch to `SUFFIX` for the name.

**Comment by drewr** [comment]
It's a little verbose, but I updated the docs to cover the default case first with a bit more explanation.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    <server>
+      <id>sonatype-nexus-staging</id>
+      <username>user</username>
+      <password>pass</password>
+    </server>
+
+Then build:
+
+    % vagrant up
+    % gradle assemble generatePomFileForJnaPublication
```
Comment (rjernst) [inline README.md]: Just `gradle assemble` should be sufficient here (assemble is set to depend on the pom generation task).

File: README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    % vagrant halt
+
+In order for OSSRH to accept the package, it must have sources and
+javadoc jars as well.  We don't have any automation here to build
+them, so simply reuse from net.java.dev/jna:
+
+    % mkdir -p tmp
```
Comment (rjernst) [inline README.md]: I would put this under `build/tmp`, instead of creating another dir that must exist in `.gitignore`.

File: build.gradle [hunk 4]
```diff
@@ -2,7 +2,19 @@
...
 group = 'org.elasticsearch'
-version = '4.4.0'
+String upstreamVersion = '4.4.0'
+
+// Rarely, modifications may need to be made to the jna jar that
+// aren't reflected in the upstream.  A suffix can be supplied to bump
+// the local version.  See readme for details.
+String versionSuffix = System.getProperty("version.suffix")
+
+if (versionSuffix != null) {
```
Comment (rjernst) [inline build.gradle]: You can avoid this conditional by using a default value of "" in the getProperty call above.

**Review by rjernst — APPROVED** [review]
LGTM. A few last nits.

File: README.md (hunk 3 above)
Comment (drewr) [reply README.md]: Even better, just used `build`.

File: README.md (hunk 2 above)
Comment (drewr) [reply README.md]: Great, didn't realize the dep there.

File: build.gradle (hunk 4 above)
Comment (drewr) [reply build.gradle]: Of course, didn't realize that version. Thanks.

**Comment by drewr** [comment]
Thanks @rjernst!
